File: scala/client/tests/conformance/Adapter.scala
package convex

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.io.PrintWriter
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.nio.charset.StandardCharsets
import java.time.Duration
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

/** Test-only NDJSON adapter protocol v1. stdout is exclusively protocol events. */
object Adapter:
  @volatile private[convex] var testRelayBeforePublish
      : (String, LiveClient.Subscription, LiveClient.Update) => Unit = (_, _, _) => ()
  @volatile private[convex] var testRelayAfterPublishAttempt
      : (String, LiveClient.Subscription) => Unit = (_, _) => ()

  def main(args: Array[String]): Unit =
    Option(System.getenv("ADAPTER_LISTEN")).filter(_.nonEmpty) match
      case None =>
        run(
          System.in,
          System.out,
          Option(System.getenv("CONVEX_URL")).getOrElse("")
        )
      case Some(address) =>
        val server = bind(address)
        try
          val controller = server.accept()
          try
            run(
              controller.getInputStream,
              controller.getOutputStream,
              Option(System.getenv("CONVEX_URL")).getOrElse("")
            )
          finally controller.close()
        finally server.close()

  private[convex] def bind(address: String): ServerSocket =
    val split = address.lastIndexOf(':')
    if split < 1 then throw new IllegalArgumentException("ADAPTER_LISTEN must be host:port")
    val server = new ServerSocket()
    server.bind(
      new InetSocketAddress(
        InetAddress.getByName(address.substring(0, split)),
        address.substring(split + 1).toInt
      )
    )
    server

  private[convex] def run(
      source: InputStream,
      sink: OutputStream,
      deployment: String
  ): Unit =
    val input = new BufferedReader(new InputStreamReader(source, StandardCharsets.UTF_8))
    val output = new Output(sink)
    val relays = new RelayRegistry
    var client: Option[ConvexClient] = None
    var live: Option[LiveClient] = None
    var stopped = false

    def ensureClient(): ConvexClient = client.getOrElse {
      if deployment.isBlank then throw new IllegalStateException("CONVEX_URL is required")
      val created = new ConvexClient(deployment)
      client = Some(created)
      created
    }

    while !stopped do
      val line = input.readLine()
      if line == null then stopped = true
      else
        var id = ""
        try
          val command = ConvexClient.json.readTree(line)
          id = command.path("id").asText("")
          command.path("op").asText("") match
            case "hello" =>
              if command.path("protocolVersion").asInt() != 1 then
                throw new IllegalArgumentException("unsupported adapter protocol version")
              output.write(
                event("ready", id)
                  .put("protocolVersion", 1)
                  .put("language", "scala")
                  .put("implementation", "native-scala-3.3.5")
                  .put("runtime", util.Properties.versionNumberString)
              )
            case "close" =>
              relays.closeAndClear()
              live.foreach(_.close())
              client.foreach(_.close())
              output.write(event("closed", id))
              stopped = true
            case "setAuth" =>
              ensureClient().setAuth(command.path("token").asText())
              output.write(event("ack", id))
            case operation @ ("query" | "mutation" | "action") =>
              val arguments =
                if command.has("args") then command.get("args")
                else ConvexClient.json.createObjectNode()
              val result = operation match
                case "query" =>
                  ensureClient().query(command.path("path").asText(), arguments)
                case "mutation" =>
                  ensureClient().mutation(command.path("path").asText(), arguments)
                case _ =>
                  ensureClient().action(command.path("path").asText(), arguments)
              output.write(resultEvent(id, result))
            case "subscribe" =>
              val subscriptionId = command.path("subscriptionId").asText()
              if subscriptionId.isBlank then
                throw new IllegalArgumentException("subscriptionId is required")
              relays.remove(subscriptionId).foreach(_.close())
              val manager = live.getOrElse {
                if deployment.isBlank then throw new IllegalStateException("CONVEX_URL is required")
                val created = new LiveClient(deployment)
                live = Some(created)
                created
              }
              val subscription = manager.subscribe(
                command.path("path").asText(),
                if command.has("args") then command.get("args")
                else ConvexClient.json.createObjectNode()
              )
              relays.registerAndWrite(subscriptionId, subscription, output, event("ack", id))
              val worker = new Thread(
                () => relay(subscriptionId, subscription, relays, output),
                s"convex-scala-adapter-$subscriptionId"
              )
              worker.setDaemon(true)
              worker.start()
            case "unsubscribe" =>
              val subscriptionId = command.path("subscriptionId").asText("")
              relays.remove(subscriptionId).foreach(_.close())
              relays.writeBarrier(output, event("ack", id))
            case "debugDisconnect" =>
              live
                .getOrElse(
                  throw new IllegalStateException("Live WebSocket is not connected")
                )
                .debugDisconnect()
              output.write(event("ack", id))
            case operation =>
              throw new IllegalArgumentException(s"unknown operation: $operation")
        catch case error: Exception => output.write(failure(id, "", error))

    relays.closeAndClear()
    live.foreach(_.close())
    client.foreach(_.close())

  private def relay(
      subscriptionId: String,
      subscription: LiveClient.Subscription,
      relays: RelayRegistry,
      output: Output
  ): Unit =
    var running = true
    while running && relays.isActive(subscriptionId, subscription) do
      try
        val update = subscription.nextUpdate(Duration.ofDays(1))
        testRelayBeforePublish(subscriptionId, subscription, update)
        relays.publishIfActive(
          subscriptionId,
          subscription,
          output,
          if update.error != null then failure("", subscriptionId, update.error)
          else subscriptionEvent(subscriptionId, update)
        )
        testRelayAfterPublishAttempt(subscriptionId, subscription)
      catch
        case error: Exception =>
          relays.publishIfActive(
            subscriptionId,
            subscription,
            output,
            failure("", subscriptionId, error)
          )
          running = false

  private[convex] def subscriptionEvent(
      subscriptionId: String,
      update: LiveClient.Update
  ): ObjectNode =
    val response = event("subscription", "").put("subscriptionId", subscriptionId)
    response.set("value", update.value)
    if update.logs.nonEmpty then
      response.set("logs", ConvexClient.json.valueToTree(update.logs.asJava))
    response

  private[convex] def resultEvent(
      id: String,
      result: ConvexClient.Result
  ): ObjectNode =
    val response = event("result", id)
    response.set("value", result.value)
    if result.logs.nonEmpty then
      response.set("logs", ConvexClient.json.valueToTree(result.logs.asJava))
    response

  private[convex] def failure(
      id: String,
      subscriptionId: String,
      error: Exception
  ): ObjectNode =
    val response = event(
      if subscriptionId.isBlank then "error" else "subscription",
      if subscriptionId.isBlank then id else ""
    )
    if subscriptionId.nonEmpty then response.put("subscriptionId", subscriptionId)
    val name = error match
      case _: ConvexClient.FunctionError  => "FunctionError"
      case _: ConvexClient.ProtocolError  => "ProtocolError"
      case _: ConvexClient.TransportError => "TransportError"
      case _                              => error.getClass.getSimpleName
    val detail = response
      .putObject("error")
      .put("name", name)
      .put("message", Option(error.getMessage).getOrElse(name))
    error match
      case function: ConvexClient.FunctionError =>
        if function.data != null then detail.set("data", function.data)
        if function.logs.nonEmpty then
          response.set("logs", ConvexClient.json.valueToTree(function.logs.asJava))
      case _ => ()
    response

  private def event(kind: String, id: String): ObjectNode =
    val result = ConvexClient.json.createObjectNode().put("type", kind)
    if id.nonEmpty then result.put("id", id)
    result

  private[convex] final class Output(sink: OutputStream):
    private val writer = new PrintWriter(sink, true, StandardCharsets.UTF_8)
    def write(value: ObjectNode): Unit = synchronized(writer.println(value))

  private[convex] final class RelayRegistry:
    private val monitor = new Object
    private val active = mutable.HashMap.empty[String, LiveClient.Subscription]
    private var closing = false

    def isActive(id: String, subscription: LiveClient.Subscription): Boolean =
      monitor.synchronized(!closing && active.get(id).contains(subscription))

    def remove(id: String): Option[LiveClient.Subscription] =
      monitor.synchronized(active.remove(id))

    def registerAndWrite(
        id: String,
        subscription: LiveClient.Subscription,
        output: Output,
        acknowledgement: ObjectNode
    ): Unit = monitor.synchronized {
      if closing then throw new IllegalStateException("adapter is closing")
      active(id) = subscription
      output.write(acknowledgement)
    }

    def writeBarrier(output: Output, acknowledgement: ObjectNode): Unit =
      monitor.synchronized(output.write(acknowledgement))

    def publishIfActive(
        id: String,
        subscription: LiveClient.Subscription,
        output: Output,
        response: ObjectNode
    ): Unit = monitor.synchronized {
      if !closing && active.get(id).contains(subscription) then output.write(response)
    }

    def closeAndClear(): List[LiveClient.Subscription] = monitor.synchronized {
      closing = true
      val subscriptions = active.values.toList
      active.clear()
      subscriptions
    }

File: clojure/client/tests/conformance/adapter.clj
(ns adapter
  "Test-only NDJSON adapter v1. stdout is exclusively protocol events."
  (:gen-class)
  (:require [clojure.data.json :as json]
            [clojure.string :as string]
            [convex.client :as convex])
  (:import [java.io BufferedInputStream ByteArrayOutputStream InputStream OutputStream]
           [java.net InetAddress InetSocketAddress ServerSocket]
           [java.nio ByteBuffer]
           [java.nio.charset CodingErrorAction StandardCharsets]
           [java.util ArrayDeque]
           [java.util.concurrent CompletableFuture ConcurrentHashMap ExecutionException TimeUnit TimeoutException]
           [java.util.concurrent.atomic AtomicLong]))

(def ^:private max-command-bytes (* 1024 1024))
(def ^:private max-output-count 16)
(def ^:private max-output-bytes (* 4 1024 1024))
(def ^:private max-close-output-ms 1000)

(def before-publish-hook (atom nil))
(def inside-publication-lock-hook (atom nil))

(defn reset-test-hooks! []
  (reset! before-publish-hook nil)
  (reset! inside-publication-lock-hook nil))

(defn- error-name [error]
  (case (:kind (ex-data error))
    :function "FunctionError"
    :protocol "ProtocolError"
    :transport "TransportError"
    :closed "ClosedError"
    "Error"))

(defn- error-event [id subscription-id error]
  (let [data (ex-data error)
        detail (cond-> {"name" (error-name error)
                        "message" (or (.getMessage ^Throwable error) "error")}
                 (contains? data :data) (assoc "data" (:data data)))]
    (cond-> {"type" (if subscription-id "subscription" "error")
             "error" detail}
      (and (nil? subscription-id) (seq id)) (assoc "id" id)
      (seq subscription-id) (assoc "subscriptionId" subscription-id)
      (seq (:logs data)) (assoc "logs" (vec (:logs data))))))

(defrecord OutputEnvelope [bytes current? before-check droppable? writing completion terminal?])

(defrecord BoundedWriter [^OutputStream output ^ArrayDeque queue bytes accepting worker]
  java.lang.AutoCloseable
  (close [_] (reset! accepting false)))

(defn- writer [output]
  (let [writer (->BoundedWriter output (ArrayDeque.) (atom 0) (atom true) (atom nil))
        worker (doto
                (Thread.
                 (fn []
                   (loop []
                     (let [envelope
                           (locking writer
                             (while (and (.isEmpty ^ArrayDeque (:queue writer)) @(:accepting writer))
                               (.wait writer))
                             (when-let [^OutputEnvelope next (.peekFirst ^ArrayDeque (:queue writer))]
                               (reset! (:writing next) true)
                               next))]
                       (when envelope
                         (try
                           (.write ^OutputStream (:output writer) ^bytes (:bytes envelope))
                           (.flush ^OutputStream (:output writer))
                           (.complete ^CompletableFuture (:completion envelope) nil)
                           (catch Throwable error
                             (.completeExceptionally ^CompletableFuture (:completion envelope) error)
                             (reset! (:accepting writer) false))
                           (finally
                             (locking writer
                               (.removeFirstOccurrence ^ArrayDeque (:queue writer) envelope)
                               (swap! (:bytes writer) - (alength ^bytes (:bytes envelope)))
                               (.notifyAll writer))))
                         (when-not (:terminal? envelope) (recur)))))))
                 (.setName "convex-clojure-adapter-writer")
                 (.setDaemon true)
                 (.start))]
    (reset! (:worker writer) worker)
    writer))

(defn- encoded-line [event]
  (.getBytes (str (json/write-str event) "\n") StandardCharsets/UTF_8))

(defn- remove-oldest-droppable! [writer]
  (when-let [oldest
             (first (filter #(and (:droppable? %) (not @(:writing %)))
                            (iterator-seq (.iterator ^ArrayDeque (:queue writer)))))]
    (.remove ^ArrayDeque (:queue writer) oldest)
    (swap! (:bytes writer) - (alength ^bytes (:bytes oldest)))
    (.complete ^CompletableFuture (:completion oldest) false)
    true))

(defn- enqueue-event! [writer event current? before-check droppable? terminal?]
  (let [bytes (encoded-line event)
        completion (CompletableFuture.)]
    (locking writer
      (when before-check (before-check))
      (if-not (and @(:accepting writer) (current?))
        (.complete completion false)
        (do
          (while (and (or (>= (.size ^ArrayDeque (:queue writer)) max-output-count)
                          (> (+ @(:bytes writer) (alength bytes)) max-output-bytes))
                      (remove-oldest-droppable! writer)))
          (if (or (> (alength bytes) max-output-bytes)
                  (>= (.size ^ArrayDeque (:queue writer)) max-output-count)
                  (> (+ @(:bytes writer) (alength bytes)) max-output-bytes))
            (if droppable?
              (.complete completion false)
              (do
                (reset! (:accepting writer) false)
                (.completeExceptionally completion
                                        (ex-info "adapter output budget exhausted"
                                                 {:kind :transport}))))
            (let [envelope (->OutputEnvelope bytes current? before-check droppable?
                                             (atom false) completion terminal?)]
              (.addLast ^ArrayDeque (:queue writer) envelope)
              (swap! (:bytes writer) + (alength bytes))
              (when terminal? (reset! (:accepting writer) false))
              (.notifyAll writer)))))
      completion)))

(defn- write-event! [writer event]
  (enqueue-event! writer event (constantly true) nil false false))

(defn- write-if-current! [writer current? before-check event]
  (enqueue-event! writer event current? before-check true false))

(defn- close-writer! [writer event]
  (let [completion (enqueue-event! writer event (constantly true) nil false true)]
    ;; Normal close waits for all prior output. A peer that stopped reading can
    ;; still be released by closing its socket; no producer holds this monitor
    ;; while the writer thread performs blocking I/O.
    (try
      (.get ^CompletableFuture completion max-close-output-ms TimeUnit/MILLISECONDS)
      (catch TimeoutException _
        (reset! (:accepting writer) false)
        false)
      (catch ExecutionException _ false))))

(defn- await-output! [writer]
  ;; EOF and fatal-input paths do not have a close event to wait on. Queue a
  ;; zero-byte barrier so tests and pipe users never observe run-adapter!
  ;; returning before its final protocol event reaches the stream.
  (let [completion (CompletableFuture.)
        barrier (->OutputEnvelope (byte-array 0) (constantly true) nil false
                                  (atom false) completion false)]
    (locking writer
      (while (and (>= (.size ^ArrayDeque (:queue writer)) max-output-count)
                  (remove-oldest-droppable! writer)))
      (if (and @(:accepting writer)
               (< (.size ^ArrayDeque (:queue writer)) max-output-count))
        (do
          (.addLast ^ArrayDeque (:queue writer) barrier)
          (.notifyAll writer))
        (.complete completion false)))
    (try
      (.get completion max-close-output-ms TimeUnit/MILLISECONDS)
      (catch TimeoutException _ false)
      (catch ExecutionException _ false))))

(defn- relay! [subscription-id registration registrations writer]
  (doto
   (Thread.
    (fn []
      (try
        (while @(:active (:subscription registration))
          (let [update (try
                         (convex/next-update (:subscription registration) 250)
                         (catch clojure.lang.ExceptionInfo error
                           (when-not (= "timed out waiting for Live update" (.getMessage error))
                             (throw error))
                           nil))]
            (when update
              (when-let [hook @before-publish-hook]
                (hook subscription-id (:generation registration) update))
              (let [event (if-let [detail (:error update)]
                            (error-event nil subscription-id
                                         (ex-info (:message detail)
                                                  {:kind (:kind detail)
                                                   :data (:data detail)
                                                   :logs (:logs update)}))
                            (cond-> {"type" "subscription"
                                     "subscriptionId" subscription-id
                                     "value" (:value update)}
                              (seq (:logs update)) (assoc "logs" (:logs update))))]
                (write-if-current!
                 writer
                 #(and (identical? registration (.get registrations subscription-id))
                       @(:active (:subscription registration)))
                 #(when-let [hook @inside-publication-lock-hook]
                    (hook subscription-id (:generation registration)))
                 event)))))
        (catch InterruptedException _ (.interrupt (Thread/currentThread)))
        (catch Throwable error
          (write-if-current!
           writer
           #(and (identical? registration (.get registrations subscription-id))
                 @(:active (:subscription registration)))
           nil
           (error-event nil subscription-id error))))))
    (.setName (str "convex-clojure-adapter-relay-" (:generation registration)))
    (.setDaemon true)
    (.start)))

(defn- ack [id] {"type" "ack" "id" id})

(defn- decode-command-line [^ByteArrayOutputStream bytes]
  (try
    (let [decoder (doto (.newDecoder StandardCharsets/UTF_8)
                    (.onMalformedInput CodingErrorAction/REPORT)
                    (.onUnmappableCharacter CodingErrorAction/REPORT))
          decoded (str (.decode decoder (ByteBuffer/wrap (.toByteArray bytes))))]
      (if (.endsWith decoded "\r")
        (subs decoded 0 (dec (count decoded)))
        decoded))
    (catch java.nio.charset.CharacterCodingException error
      (throw (ex-info "adapter command is not valid UTF-8"
                      {:kind :protocol :cause error})))))

(defn- read-command-line! [^InputStream input]
  ;; BufferedReader.readLine grows until newline. A hostile controller could
  ;; therefore allocate tens of megabytes before JSON decoding gets a chance to
  ;; reject anything. Count encoded bytes first and fail the connection at 1 MiB.
  (let [bytes (ByteArrayOutputStream. 8192)]
    (loop []
      (let [byte (.read input)]
        (cond
          (= -1 byte) (when (pos? (.size bytes)) (decode-command-line bytes))
          (= 10 byte) (decode-command-line bytes)
          (>= (.size bytes) max-command-bytes)
          (throw (ex-info (str "adapter command exceeds " max-command-bytes " encoded bytes")
                          {:kind :protocol :fatal true}))
          :else (do (.write bytes byte) (recur)))))))

(defn run-adapter! [input output deployment]
  (let [reader (BufferedInputStream. input 8192)
        out (writer output)
        client (atom nil)
        registrations (ConcurrentHashMap.)
        generation (AtomicLong. 1)]
    (letfn [(get-client []
              (or @client
                  (let [url (or deployment
                                (throw (ex-info "CONVEX_URL is required" {:kind :protocol})))
                        created (convex/client url)]
                    (when-let [token (System/getenv "CONVEX_AUTH_TOKEN")]
                      (when (seq token) (convex/set-auth! created token)))
                    (reset! client created)
                    created)))
            (process! [command]
              (let [id (get command "id") operation (get command "op")]
                (case operation
                  "hello"
                  (do
                    (when-not (= 1 (get command "protocolVersion"))
                      (throw (ex-info "unsupported adapter protocol version" {:kind :protocol})))
                    (write-event! out {"type" "ready" "id" id "protocolVersion" 1
                                       "language" "clojure" "implementation" "native-clojure-jdk21"
                                       "runtime" (System/getProperty "java.runtime.version")})
                    false)

                  ("query" "mutation" "action")
                  (let [result (convex/call (get-client) operation (get command "path")
                                            (get command "args" {}))]
                    (write-event! out (cond-> {"type" "result" "id" id "value" (:value result)}
                                        (seq (:logs result)) (assoc "logs" (:logs result))))
                    false)

                  "setAuth"
                  (do (convex/set-auth! (get-client) (get command "token" ""))
                      (write-event! out (ack id)) false)

                  "subscribe"
                  (let [subscription-id (get command "subscriptionId")
                        _ (when-not (seq subscription-id)
                            (throw (ex-info "subscriptionId is required" {:kind :protocol})))
                        old (.remove registrations subscription-id)]
                    ;; Closing the old generation before publishing ack is the replacement barrier.
                    (when old (.close ^java.lang.AutoCloseable (:subscription old)))
                    (let [subscription (convex/subscribe-client (get-client) (get command "path")
                                                                (get command "args" {}))
                          registration {:generation (.getAndIncrement generation)
                                        :subscription subscription}]
                      (.put registrations subscription-id registration)
                      (write-event! out (ack id))
                      (relay! subscription-id registration registrations out))
                    false)

                  "unsubscribe"
                  (do
                    (when-let [registration (.remove registrations (get command "subscriptionId"))]
                      (.close ^java.lang.AutoCloseable (:subscription registration)))
                    ;; The writer lock is shared with relay publication. Once ack is visible,
                    ;; no invalidated generation can publish another event.
                    (write-event! out (ack id))
                    false)

                  "debugDisconnect"
                  (do (convex/debug-disconnect-client! (get-client))
                      (write-event! out (ack id)) false)

                  "close"
                  (do
                    (doseq [registration (iterator-seq (.iterator (.values registrations)))]
                      (.close ^java.lang.AutoCloseable (:subscription registration)))
                    (.clear registrations)
                    (when @client (.close ^java.lang.AutoCloseable @client))
                    (close-writer! out {"type" "closed" "id" id})
                    true)

                  (throw (ex-info (str "unknown operation " operation) {:kind :protocol})))))]
      (try
        (loop []
          (let [line (try
                       (read-command-line! reader)
                       (catch clojure.lang.ExceptionInfo error
                         (write-event! out (error-event nil nil error))
                         ::fatal-input))]
            (cond
              (nil? line) nil
              (= ::fatal-input line) nil
              (string/blank? line) (recur)
              :else
              (let [command (try
                              (json/read-str line)
                              (catch Exception error
                                (write-event! out
                                              (error-event nil nil
                                                           (ex-info (str "decode command: " (.getMessage error))
                                                                    {:kind :protocol})))
                                nil))
                    stop? (if command
                            (try
                              (process! command)
                              (catch Throwable error
                                ;; Per-command failures retain correlation and the stream continues.
                                (write-event! out (error-event (get command "id") nil error))
                                false))
                            false)]
                (when-not stop? (recur))))))
        (finally
          (doseq [registration (iterator-seq (.iterator (.values registrations)))]
            (.close ^java.lang.AutoCloseable (:subscription registration)))
          (.clear registrations)
          (when @client (.close ^java.lang.AutoCloseable @client))
          (await-output! out))))))

(defn parse-listen-address [address]
  (let [split (.lastIndexOf ^String address ":")]
    (when (or (<= split 0) (= split (dec (count address))))
      (throw (IllegalArgumentException. "ADAPTER_LISTEN must be host:port")))
    [(subs address 0 split) (Integer/parseInt (subs address (inc split)))]))

(defn- bind [address]
  (let [[host port] (parse-listen-address address)
        server (ServerSocket.)]
    (.bind server (InetSocketAddress. (InetAddress/getByName host) port))
    server))

(defn -main [& _]
  (if-let [address (System/getenv "ADAPTER_LISTEN")]
    (with-open [server (bind address)
                socket (.accept server)]
      (run-adapter! (.getInputStream socket) (.getOutputStream socket)
                    (System/getenv "CONVEX_URL")))
    (run-adapter! System/in System/out (System/getenv "CONVEX_URL"))))
